api_handler/api_calls.py

```diff
@@ -0,0 +1,87 @@
+import json, logging
+from json.decoder import JSONDecodeError
+from typing import Any, Dict, Union, Optional
+from constants import (
+    CANVAS_SCOPE
+)
+
+# third-party libraries
+from requests import Response
+from umich_api.api_utils import ApiUtil
+
+logger = logging.getLogger(__name__)
+class APIHandler:
+    def __init__(self, props: Dict[str, Optional[str]]) -> None:
+        self.api_handler: ApiUtil = ApiUtil(props.get('api_url'), props.get('api_client'), props.get('api_secret'))
+        self.max_req_attempts: int = int(props.get('retry_attempts')) #type: ignore
+    
+    def check_if_response_successful(self, response: Response) -> bool:
+        """
+        Checks whether response has 200 status code and the JSON text can be parsed.
+
+        :param response: Response from ApiUtil.api_call
+        :type response: Response
+        :return: True or False depending on whether the response was successful
+        :rtype: bool
+        """
+        response_successful: bool = True
+
+        status_code: int = response.status_code
+        if not response.ok:
+            logger.error(f'Received irregular status code: {status_code}')
+            response_successful = False
+        else:
+            try:
+                json.loads(response.text)
+            except JSONDecodeError:
+                logger.error('JSONDecodeError encountered')
+                response_successful = False
+
+        if not response_successful:
+            logger.error(response.text)
+        return response_successful
+    
+    def get_next_page(self, response: Response) -> Any:
+        return self.api_handler.get_next_page(response)
+
+    def api_call_with_retries(self,
+        url: str,
+        method: str,
+        payload: Union[Dict[str, Any], None] = None,
+    ) -> Any:
+        """
+        Pulls data from the UM API Directory, handling errors and retrying if necessary.
+
+        When the maximum number of request attempts is reached, the function logs an error and returns None.
+        :param url: URL ending for request
+        :type url: string
+        :param method: Request method that should be used (e.g. "GET", "PUT")
+        :type method: string
+        :param payload: Dictionary to include in the request body
+        :type payload: Dictionary with string keys or None, optional
+        :return: Either a Response object or None
+        :rtype: Response or None
+        """
+        if payload is None:
+            request_payload = dict()
+        else:
+            request_payload = payload
+
+        logger.debug('Making a request for data...')
+
+        for i in range(1, self.max_req_attempts + 1):
```

Comment:
I think this is fine to have here for now, but I [created an issue on api-utils](https://github.com/tl-its-umich-edu/api-utils-python/issues/40) where we could just add this support for retries to the library.

Reply:
Yeah... this feature  has been in back burner for longest. I don't know when this gets done.